src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,774 @@
+
+    if (subscan_res.has_failure()) {
+        // Failed to scan the beginning of the offset range.
+        co_return subscan_res.as_failure();
+    }
+
+    vlog(
+      _ctxlog.debug,
+      "Left scan result, done: {}, offset range: {}-{}, file offset: "
+      "{}, timestamps: {}-{}",
+      left_scan.done,
+      left_scan.base_offset,
+      left_scan.last_offset,
+      left_scan.filepos,
+      left_scan.first_timestamp,
+      left_scan.max_timestamp);
+
+    if (!subscan_res.value().done) {
+        // Subscan failed to find the data timestamp (this usually
+        // requires some scanning past the target)
+        co_return make_error_code(error_outcome::scan_failed);
+    }
+    left_scan = subscan_res.value();
+    co_return left_scan;
+}
+
+ss::future<result<segment_upload::subscan_res>> segment_upload::subscan_right(
+  std::vector<ss::lw_shared_ptr<storage::segment>> segments,
+  model::offset range_last,
+  model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    subscan_res right_scan;
+    // The left and right might point to the same segment if we
+    // have only one segment in the list.
+    auto segment = segments.back();
+    auto ix_last = segment->index().find_nearest(range_last);
+    model::offset target;
+    if (!ix_last.has_value()) {
+        // Fallback to full segment scan
+        target = segment->offsets().base_offset;
+    } else {
+        target = ix_last.value().offset;
+    }
+    auto subscan_res = co_await subscan(
+      _part,
+      inclusive_offset_range(target, range_last),
+      ix_last->filepos,
+      true,
+      deadline);
+
+    if (subscan_res.has_failure()) {
+        // Failed to scan the beginning of the offset range.
+        co_return subscan_res.as_failure();
+    }
+
+    vlog(
+      _ctxlog.debug,
+      "Right scan result, done: {}, offset range: {}-{}, file offset: "
+      "{}, "
+      "timestamps: {}-{}",
+      right_scan.done,
+      right_scan.base_offset,
+      right_scan.last_offset,
+      right_scan.filepos,
+      right_scan.first_timestamp,
+      right_scan.max_timestamp);
+
+    if (!subscan_res.value().done) {
+        // Subscan failed to find the data timestamp (this usually
+        // requires some scanning past the target)
+        co_return make_error_code(error_outcome::scan_failed);
+    }
+    right_scan = subscan_res.value();
+    co_return right_scan;
+}
+
+ss::future<result<cloud_upload_parameters>>
+segment_upload::compute_upload_parameters(
+  inclusive_offset_range range, model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    try {
+        auto segments = [&] {
+            auto log = _part->log();
+            auto dli = dynamic_cast<const storage::disk_log_impl*>(log.get());
+            auto base_it = dli->segments().lower_bound(range.base);
+            std::vector<ss::lw_shared_ptr<storage::segment>> segments;
+            for (auto it = base_it; it != dli->segments().end(); it++) {
```

Comment:
I was under the impression that we are removing the `log::segments()` abstraction.